src/main/java/io/spokestack/spokestack/wakeword/RingBuffer.java

```diff
@@ -107,6 +107,10 @@ public void write(float value) {
     }
 
     private int pos(int x) {
-        return Math.floorMod(x, this.data.length);
+        int mod = x % this.data.length;
```

Comment:
not quite that simple, you want knuth's floored division: https://en.wikipedia.org/wiki/Modulo_operation
see how it is implemented in openjdk: http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/lang/Math.java#l1138

for example, floormod(-2, 10) should be 8

Reply:
while this is probably fine for any real usage of the ring buffer, it is still an incorrect implementation of floormod
consider `seek(-this.data.length - 2)`
it would be better just to port the openjdk floormod implementation in favor of this ad hoc solution

Reply:
As usual, my instinct was to save a couple cycles by avoiding the operations we didn't strictly need. I wouldn't actually call my version `floorMod` since it relies on internal knowledge of the buffer, but point taken.

Reply:
this code is still still using the `%` operator here, and java's `%` operator has nothing to do with modular arithmetic
what I was saying was to leave this function alone (aside from removing the `Math.` prefix) and just copy the openjdk floorMod/floorDiv functions into this module
this way, the `pos` function does its job of computing the physical offset from a logical offset, and floorMod does its job of computing the modulus operator

Reply:
This literally is the openjdk implementation of floorMod(int, int), just with `this.data.length` substituted for `y`. Their javadoc specifies the _relationship_ between floorMod and floorDiv, but the implementation doesn't actually call floorDiv.

I considered putting it in its own method, but that would be adding function call overhead for a questionable gain in clarity. If you think it's worth it, though, I guess the JVM might inline it at runtime.

Reply:
huh, I must have been looking at an older implementation
I retract my comment, then, lgtm